lisa/features/gpu.py

```diff
@@ -0,0 +1,125 @@
+
+from lisa.feature import Feature
+from lisa.operating_system import Linux, Redhat, Ubuntu
+from lisa.tools import Uname, Wget
+from lisa.util import LisaException
+
+FEATURE_NAME_GPU = "Gpu"
+
+
+class compute_sdk(Enum):
+    GRID = 0
+    CUDA = 1
+
+
+class Gpu(Feature):
+    def __init__(self, node: Any, platform: Any) -> None:
+        super().__init__(node, platform)
+        self._log = self._node.log
+
+    @classmethod
+    def name(cls) -> str:
+        return FEATURE_NAME_GPU
+
+    def _install_grid_driver(self, version: str) -> None:
+        self._log.info("Starting GRID driver installation")
+
+    def _install_cuda_driver(self, version: str = "10.1.105-1") -> None:
+        self._log.info("Starting CUDA driver installation")
+        cuda_repo = ""
+        linux_os: Linux = cast(Linux, self._node.os)
+
+        # CUDA driver installation for redhat distros
+        if isinstance(self._node.os, Redhat):
+            cuda_repo_pkg = f"cuda-repo-rhel7-{version}.x86_64.rpm"
+            cuda_repo = (
+                "http://developer.download.nvidia.com/"
+                f"compute/cuda/repos/rhel7/x86_64/{cuda_repo_pkg}"
+            )
+            linux_os = Redhat(self._node)
+
+        # CUDA driver installation for Ubuntu distros
+        elif isinstance(self._node.os, Ubuntu):
+            os_version = self._node.os.get_os_version()
+            cuda_repo_pkg = f"cuda-repo-ubuntu{os_version}//./_{version}_amd64.deb"
+            cuda_repo = (
+                "http://developer.download.nvidia.com/compute/"
+                f"cuda/repos/ubuntu{os_version}//.//x86_64/${cuda_repo_pkg}"
+            )
+            linux_os = Ubuntu(self._node)
+
+        else:
+            raise LisaException(
+                f"Distro {self._node.os._get_detect_string(self._node)}"
+                "not supported to install CUDA driver."
+            )
+
+        wget_tool = self._node.tools[Wget]
+        # download the cuda driver at /tmp/
+        wget_tool.get(cuda_repo, "/tmp/", cuda_repo_pkg)
+        # install the cuda driver rpm
+        install_result = linux_os.install_packages(
+            f"/tmp/{cuda_repo_pkg}", signed=False
+        )
+        if install_result.exit_code != 0:
+            raise LisaException(
+                f"Failed to install {cuda_repo_pkg}. stdout: {install_result.stdout}"
+            )
+        else:
+            self._log.info("Sucessfully installed cuda-drivers")
+
+    def install_gpu_dep(self) -> None:
+        uname_tool = self._node.tools[Uname]
+        uname_ver = uname_tool.get_linux_information().uname_version
+
+        # install dependency libraries for redhat and CentOS
+        if isinstance(self._node.os, Redhat):
+            # install the kernel-devel and kernel-header packages
+            package_name = f"kernel-devel-{uname_ver} kernel-headers-{uname_ver}"
+            install_result = self._node.os.install_packages(package_name)
+            if install_result.exit_code != 0:
```

Comment:
Let's move this check into install_packages as discussed.

Reply:
> Let's move this check into install_packages as discussed.

What we discussed is that we should be able to return the output/error from install_packages. But I believe handling on error and stuff should be left to caller of install_packages?

Reply:
Does the `exit_code == 0` mean success? If so, it can be in the install_packages method. If some installation failure is expected, we can add one more parameter like `ignore_error`.

Reply:
> Does the `exit_code == 0` mean success? If so, it can be in the install_packages method. If some installation failure is expected, we can add one more parameter like `ignore_error`.

I think it will be more appropriate if the caller method has the full knowledge of cmd output/error and then can take action accordingly. It might want to ignore particular kind of stderr and not others?

Reply:
The command is installation here, and the exact command relays on implementation of `install_packages`. So it makes sense more to check the results in `install_packages`.

Reply:
> The command is installation here, and the exact command relays on implementation of `install_packages`. So it makes sense more to check the results in `install_packages`.

Done